Fixed error when receiving null $argv in Reflection::reflectClass/reflectFunction

mtorromeo opened the pull request:
…
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Given the loose typehints in `ReflectionClass` and `ReflectionMethod` it can happen that they receive `null` as the `$argv` parameter, leading to an error.

In practice this happened to me when using laminas-json-server which contains the following code with `$argv` defaulting to `null`:
```php
    public function setClass($class, $namespace = '', $argv = null)
    {
        if (2 < func_num_args()) {
            $argv = func_get_args();
            $argv = array_slice($argv, 2);
        }

        $reflection = Reflection::reflectClass($class, $argv, $namespace);

        foreach ($reflection->getMethods() as $method) {
            $definition = $this->_buildSignature($method, $class);
            $this->addMethodServiceMap($definition);
        }

        return $this;
    }
```

The code from `Laminas\Json\Server` contains its own bug given that `if (2 < func_num_args()) {` should more likely be `if (3 < func_num_args()) {`, but still it seems right to at least protect from such a situation in `laminas-server` as well.

mtorromeo committed:
Fixed error when receiving null $argv in Reflection::reflectClass/reflectFunction

Signed-off-by: Massimiliano Torromeo <massimiliano.torromeo@gmail.com>

goksiii commented:
hello @mtorromeo,

when do you plan to merge this PR because we have problems with Laminas Server?

mtorromeo commented:
@goksiii I'm not a maintainer/developer of the project and I can't merge anything. I want this merged as much as you do.

weierophinney commented:
I'll try and review today, @goksiii.

weierophinney committed:
qa: tighten up validations of $argv parameter of Reflection:reflectFunction

The documented value types were `false` or an array.
However, nothing prevented other falsy values from slipping through.
This patch tightens the logic, and expands the allowed values to include `null`, per the patch description.
Further, it adds tests to verify.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: reassign $argv early in `ReflectionClass`

Re-assigns `$argv` before looping over methods, so it doesn't need to be done repeatedly.
Adds tests to validate that non-array `$argv` values are cast to an empty array.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Adds CHANGELOG entry for #19

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
